src/solvers/imex_ark_new.jl

```diff
@@ -0,0 +1,264 @@
+        ΔU_imp,
+        prev_ΔtT_lim_to_Δu_coefs,
+        prev_ΔtT_exp_to_Δu_coefs,
+        prev_z_ΔtT_imp_to_Δu_coefs,
+        prev_nz_ΔU_imp_to_Δu_coefs,
+    )
+        t_exp = t + Δt * c_exp[stage]
+        t_imp = t + Δt * c_imp[stage]
+        Δtγ = Δt * a_imp[stage, stage]
+
+        Δu_over_Δt_from_prev_T_lim = broadcasted_dot(prev_ΔtT_lim_to_Δu_coefs, T_lim_by_stage)
+        Δu_over_Δt_from_prev_T_exp = broadcasted_dot(prev_ΔtT_exp_to_Δu_coefs, T_exp_by_stage)
+        Δu_over_Δt_from_prev_T_imp = broadcasted_dot(prev_z_ΔtT_imp_to_Δu_coefs, T_imp_by_stage)
+        Δu_from_prev_ΔU_imp = broadcasted_dot(prev_nz_ΔU_imp_to_Δu_coefs, ΔU_imp_by_stage)
+
+        if isempty(prev_ΔtT_lim_to_Δu_coefs)
+            u_plus_Δu_lim = u
+        else
+            u_plus_Δu_lim = temp_value1
+            @. u_plus_Δu_lim = u + Δt * Δu_over_Δt_from_prev_T_lim
+            lim!(u_plus_Δu_lim, p, t_exp, u)
+        end
+
+        if (
+            isempty(prev_ΔtT_exp_to_Δu_coefs) &&
+            isempty(prev_z_ΔtT_imp_to_Δu_coefs) &&
+            isempty(prev_nz_ΔU_imp_to_Δu_coefs)
+        )
+            U_before_solve = u_plus_Δu_lim
+        else
+            U_before_solve = temp_value2
+            @. U_before_solve =
+                u_plus_Δu_lim + Δt * (Δu_over_Δt_from_prev_T_exp + Δu_over_Δt_from_prev_T_imp) + Δu_from_prev_ΔU_imp
+        end
+
+        post_explicit!(U_before_solve, p, t_imp)
+
+        if !isnothing(ΔU_imp)
+            U = ΔU_imp # Use ΔU_imp as additional temporary storage.
+            @. U = U_before_solve
+
+            # Solve U ≈ U_before_solve + Δtγ * T_imp(U, p, t_imp) for U.
+            set_residual! = (residual, U) -> begin
+                T_imp!(residual, U, p, t_imp)
+                @. residual = U_before_solve - U + Δtγ * residual
+            end
+            set_jacobian! = (jacobian, U) -> T_imp!.Wfact(jacobian, U, p, Δtγ, t_imp)
+            post_implicit! = U -> post_implicit!(U, p, t_imp)
+            solve_newton!(
+                newtons_method,
+                newtons_method_cache,
+                U,
+                set_residual!,
+                set_jacobian!,
+                post_implicit!,
+                post_implicit!,
+            )
+        else
+            U = U_before_solve # There is no solve on this stage.
+        end
+
+        if !isnothing(T_lim) || !isnothing(T_exp)
+            if !isnothing(T_exp_T_lim!)
+                T_exp_T_lim!(T_exp, T_lim, U, p, t_exp)
+                dss!(T_lim, p, t_exp)
+                dss!(T_exp, p, t_exp)
+            end
+            if !isnothing(T_lim!)
+                T_lim!(T_lim, U, p, t_exp)
+                dss!(T_lim, p, t_exp)
+            end
+            if !isnothing(T_exp!)
+                T_exp!(T_exp, U, p, t_exp)
+                dss!(T_exp, p, t_exp)
+            end
+            # TODO: Can we just use T_exp_T_lim!, and not 3 different functions?
+            # TODO: Fuse the DSS calls above into a single operation.
```

Comment:
Yeah, I left `T_lim!` and `T_exp!` for backward compatibility (also, they have somewhat different interfaces / storage).

We will definitely want to fuse dss calls, in fact `T_exp_T_lim!` perform dss calls on `T_exp` and `T_lim` in ClimaAtmos, so I think we can just eliminate the dss calls here.

Reply:
Ah, I actually forgot something here that prevents us from fully eliminating the DSS calls: we should skip the DSS on the last stage with an explicit/limited tendency, since we DSS the final state at the end of each timestep. But at least we can simplify the 3 branches to 1.